### PR TITLE
Add text generation sample to example-ptb

### PR DESCRIPTION
**Pull request by hamatz**
Current example doesn't show how to use the leaning result.
So, I added sample script for the clarification.

### CODE REVIEW

File: examples/ptb/genetxt.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+This code is ported from following implementation.
+https://github.com/longjie/chainer-char-rnn/blob/master/sample.py
+
+"""
+import argparse
+import codecs
+
+from chainer import cuda, Variable
+import cPickle as pickle
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use `import six.moves.cPickle as pickle`

File: examples/ptb/genetxt.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+import chainer.functions as F
+import chainer.links as L
+from chainer import serializers
+
+import net
+import numpy as np
+
+import sys
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Move `import sys` after `import codecs`.

File: examples/ptb/genetxt.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+from chainer import cuda, Variable
+import cPickle as pickle
+
+import chainer.functions as F
+import chainer.links as L
+from chainer import serializers
+
+import net
+import numpy as np
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Please move `import numpy` after `import pickle`.

File: examples/ptb/genetxt.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+""" Sample script of recurrent neural network language model for generating text
+
+This code is ported from following implementation.
+https://github.com/longjie/chainer-char-rnn/blob/master/sample.py
+
+"""
+import argparse
+import codecs
+
+from chainer import cuda, Variable
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Don't import class `Variable`. And, move this line before `import chainer.functions`

File: examples/ptb/train_ptb.py [hunk 5]
```diff
@@ -14,6 +14,8 @@
 import numpy as np
 import six
 
+import cPickle as pickle
```
Comment (unnonouno) [inline examples/ptb/train_ptb.py]: `import six.moves.cPickle as pickle`

File: examples/ptb/genetxt.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    
+if len(primetext) > 0:
+    prev_word = Variable(np.ones((1,)).astype(np.int32) * vocab[primetext])
+    if args.gpu >= 0:
+        prev_word = cuda.to_gpu(prev_word)
+
+    prob = F.softmax(model.predictor(prev_word))     
+    sys.stdout.write(primetext + " ") 
+
+for i in xrange(args.length):
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use `six.moves.range`

File: examples/ptb/genetxt.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+if args.gpu >= 0:
+    cuda.init()
+    model.to_gpu()
+
+model.predictor.reset_state()
+
+sys.stdout = codecs.getwriter('utf_8')(sys.stdout)
+
+global prev_word
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Is this line required?

File: examples/ptb/genetxt.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+primetext = unicode(args.primetext, 'utf-8')
+    
+if len(primetext) > 0:
+    prev_word = Variable(np.ones((1,)).astype(np.int32) * vocab[primetext])
+    if args.gpu >= 0:
+        prev_word = cuda.to_gpu(prev_word)
+
+    prob = F.softmax(model.predictor(prev_word))     
+    sys.stdout.write(primetext + " ")
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use single quotes.

File: examples/ptb/genetxt.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    prob = F.softmax(model.predictor(prev_word))
+    
+    if args.sample > 0:
+        probability = cuda.to_cpu(prob.data)[0].astype(np.float64)
+        probability /= np.sum(probability)
+        index = np.random.choice(range(len(probability)), p=probability)
+    else:
+        index = np.argmax(cuda.to_cpu(prob.data))
+
+    if ivocab[index] == "<eos>":
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use single quotes.

File: examples/ptb/train_ptb.py [hunk 10]
```diff
@@ -77,6 +79,7 @@ def load_data(filename):
     test_data = test_data[:100]
 
 print('#vocab =', len(vocab))
+pickle.dump(vocab, open('vocab.bin', 'wb'))
```
Comment (unnonouno) [inline examples/ptb/train_ptb.py]: Use `with open`.

File: examples/ptb/genetxt.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+parser.add_argument('--length',     type=int,   default=20, 
+                    help='length of the generated text')
+parser.add_argument('--gpu',        type=int,   default=-1, 
+                    help='GPU ID (negative value indicates CPU)')
+args = parser.parse_args()
+
+np.random.seed(args.seed)
+
+# load vocabulary
+vocab = pickle.load(open(args.vocabulary, 'rb'))
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use `with open`.

File: examples/ptb/genetxt.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        index = np.random.choice(range(len(probability)), p=probability)
+    else:
+        index = np.argmax(cuda.to_cpu(prob.data))
+
+    if ivocab[index] == "<eos>":
+        sys.stdout.write(".")
+    else:
+        sys.stdout.write(ivocab[index] + " ")        
+
+    prev_word = Variable(np.ones((1,)).astype(np.int32) * vocab[ivocab[index]])
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: `np.array([vocab[ivocab[index]]], dtype=np.int32)` ?

**Comment by unnonouno** [comment]
Thank you for your contribution! I wrote some comments. Please check if it works on python 3.

**Comment by unnonouno** [comment]
Please chmod 755.

**Comment by hamatz** [comment]
So sorry for the inconvenience.
I've fixed the code style and some compatible issues (Python2 / 3).

Hope this works.

File: examples/ptb/genetxt.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    https://github.com/longjie/chainer-char-rnn/blob/master/sample.py
+    
+    """
+import argparse
+import codecs
+import sys
+
+import six
+import six.moves.cPickle as pickle
+import numpy as np
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Please sort alphabetical order. I mean:
```
import numpy as np
import six
import six.moves.cPickle as pickle
```

File: examples/ptb/train_ptb.py [hunk 14]
```diff
@@ -77,6 +79,8 @@ def load_data(filename):
     test_data = test_data[:100]
 
 print('#vocab =', len(vocab))
+with open("vocab.bin", "wb") as f:
```
Comment (unnonouno) [inline examples/ptb/train_ptb.py]: Use single quote

**Comment by unnonouno** [comment]
Please use `hacking` to check your coding style and pass Travis CI.
http://docs.chainer.org/en/stable/contribution.html#coding-guidelines

**Comment by hamatz** [comment]
Although I've fixed locally by using flake8 command, it seems there is still some problem..
Travis CI mentioned 'E821 "unicode" is undefined on Python3'.

I use six.PY2 for finding which version is used.
So, the line which includes 'unicode' related process is never used on Python3.
However, my code cannot pass Travis CI.

Are we not allowed to use the function like six.PY2 ?

**Comment by unnonouno** [comment]
How about like this?
```
primetext = args.primetext
if isinstance(primetext, six.binary_type):
  six.text_type(primetext, 'utf-8')
```

**Comment by okuta** [comment]
How about like this?
```python
primetext = args.primetext
if isinstance(primetext, six.binary_type):
  primetext = primetext.decode('utf-8')
```

**Comment by hamatz** [comment]
Thank you for your advice!
5/6 passed on Travis CI, it seems.

Remaining error mentioned 

" No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself. "

May I ask ... Is this my issue?

File: examples/ptb/genetxt.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    model.to_gpu()
+
+model.predictor.reset_state()
+
+primetext = args.primetext
+if isinstance(primetext, six.binary_type):
+    primetext = primetext.decode('utf-8')
+
+if len(primetext) > 0:
+    prev_word = Variable(np.ones((1,)).astype(np.int32) * vocab[primetext])
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: ``prev_word = Variable(np.array([vocab[primetext]], np.int32))``

Comment (unnonouno) [inline examples/ptb/genetxt.py]: And, please check if `primetext in vocab`.

File: examples/ptb/genetxt.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    cuda.init()
+    model.to_gpu()
+
+model.predictor.reset_state()
+
+primetext = args.primetext
+if isinstance(primetext, six.binary_type):
+    primetext = primetext.decode('utf-8')
+
+if len(primetext) > 0:
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: This example causes an error when `len(primetext) == 0`, doesn't it?

File: examples/ptb/genetxt.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+# shoud be same as n_units , described in train_ptb.py
+n_units = args.unit
+
+lm = net.RNNLM(len(vocab), n_units, train=False)
+model = L.Classifier(lm)
+
+serializers.load_npz(args.model, model)
+
+if args.gpu >= 0:
+    cuda.init()
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: `cuda.init()` is deprecated, and you need not call it.

Comment (unnonouno) [inline examples/ptb/genetxt.py]: Call `cuda.get_device(args.gpu).use()`. See `train_ptb.py`.

File: examples/ptb/genetxt.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+model.predictor.reset_state()
+
+primetext = args.primetext
+if isinstance(primetext, six.binary_type):
+    primetext = primetext.decode('utf-8')
+
+if len(primetext) > 0:
+    prev_word = Variable(np.ones((1,)).astype(np.int32) * vocab[primetext])
+    if args.gpu >= 0:
+        prev_word = cuda.to_gpu(prev_word)
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use `xp` instead of calling `to_gpu`. See `train_ptb.py`.

File: examples/ptb/genetxt.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        index = np.argmax(cuda.to_cpu(prob.data))
+
+    if ivocab[index] == '<eos>':
+        sys.stdout.write('.')
+    else:
+        sys.stdout.write(ivocab[index] + ' ')
+
+    prev_word = Variable(np.array([index], dtype=np.int32))
+    if args.gpu >= 0:
+        prev_word = cuda.to_gpu(prev_word)
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Don't use `to_gpu`.

File: examples/ptb/genetxt.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        index = np.random.choice(range(len(probability)), p=probability)
+    else:
+        index = np.argmax(cuda.to_cpu(prob.data))
+
+    if ivocab[index] == '<eos>':
+        sys.stdout.write('.')
+    else:
+        sys.stdout.write(ivocab[index] + ' ')
+
+    prev_word = Variable(np.array([index], dtype=np.int32))
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Use `xp`

**Comment by unnonouno** [comment]
Please check if this example can work on GPU.

**Comment by hamatz** [comment]
Thank you for the reviewing!
Now, this example can work on GPU.

File: examples/ptb/genetxt.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+        index = np.random.choice(range(len(probability)), p=probability)
+    else:
+        index = np.argmax(cuda.to_cpu(prob.data))
+
+    if ivocab[index] == '<eos>':
+        sys.stdout.write('.')
+    else:
+        sys.stdout.write(ivocab[index] + ' ')
+
+    prev_word = Variable(xp.array([index], dtype=xp.int32))
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Please print end of line at the end of the program.
`sys.stdout.write('\n')`

**Comment by unnonouno** [comment]
Great! It works. I wrote the last comment. I'll merge this PR after you fix it.

**Comment by hamatz** [comment]
I've done.
Thank you for the all of your effort related to this PR!

File: examples/ptb/genetxt.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+sys.stdout.write('¥n')
```
Comment (unnonouno) [inline examples/ptb/genetxt.py]: Don't use EN-mark. Use backslash.

**Comment by hamatz** [comment]
oops!
So sorry for the inconvenience..

I've fixed it.

**Comment by unnonouno** [comment]
OK! Great work. It looks good for me. Thank you very much.

**Pull request merged**